utilities/simulator_cache/cache_simulator.cc

```diff
@@ -4,42 +4,177 @@
 //  (found in the LICENSE.Apache file in the root directory).
 
 #include "utilities/simulator_cache/cache_simulator.h"
+#include "db/dbformat.h"
 
 namespace rocksdb {
-CacheSimulator::CacheSimulator(std::shared_ptr<SimCache> sim_cache)
+
+namespace {
+const std::string kGhostCachePrefix = "ghost_";
+}
+
+GhostCache::GhostCache(std::shared_ptr<Cache> sim_cache)
     : sim_cache_(sim_cache) {}
 
+bool GhostCache::Admit(const Slice& lookup_key) {
+  auto handle = sim_cache_->Lookup(lookup_key);
+  if (handle != nullptr) {
+    sim_cache_->Release(handle);
+    return true;
+  }
+  sim_cache_->Insert(lookup_key, /*value=*/nullptr, lookup_key.size(),
+                     /*deleter=*/nullptr, /*handle=*/nullptr);
+  return false;
+}
+
+CacheSimulator::CacheSimulator(std::unique_ptr<GhostCache>&& ghost_cache,
+                               std::shared_ptr<Cache> sim_cache)
+    : ghost_cache_(std::move(ghost_cache)), sim_cache_(sim_cache) {}
+
 void CacheSimulator::Access(const BlockCacheTraceRecord& access) {
+  bool admit = true;
+  const bool is_user_access =
+      BlockCacheTraceHelper::IsUserAccess(access.caller);
+  bool is_cache_miss = true;
+  if (ghost_cache_ && access.no_insert == Boolean::kFalse) {
+    admit = ghost_cache_->Admit(access.block_key);
+  }
   auto handle = sim_cache_->Lookup(access.block_key);
-  if (handle == nullptr && !access.no_insert) {
-    sim_cache_->Insert(access.block_key, /*value=*/nullptr, access.block_size,
-                       /*deleter=*/nullptr, /*handle=*/nullptr);
+  if (handle != nullptr) {
+    sim_cache_->Release(handle);
```

Comment:
Can you remind me of why release the handle upon a hit?

Reply:
The reason is that each entry maintains a ref counter. It is incremented by one inside the lookup so we need to call Release to decrement the ref count.